Refactor/tests

akaviaLab opened the pull request:
* [X] fix #974 
* [X] description of feature/fix
Move the test suite from src/cobra/test to src/tests. Also removed __init__.py from the test suite.
* [X] add an entry to the [next release](../release-notes/next-release.md)
Moved the test suite out of the source tree

Looks like this broke the test suite. Odd. Tox worked on my computer.

Commit:
moved tests to src directory

Commit:
moved tests to src directory

Commit:
tests no longer use __init__.py

Commit:
turned compare_models to a function fixture

Commit:
update conftest.py to Python 3.6

Commit:
fixed minor typing error

Commit:
updated release notes for next-release.md

synchon commented:
I believe what @Midnighter meant in #974 was to move `src/cobra/test/` to `tests/` (or, `test/`). Personally, I would not do this until we have completed the codebase modernization and `flake8` is running in CI/CD. As this move will migrate data files and the tests, this will also need to take care of `cobra.test.create_test_model()`. Maybe others can also add their opinion about the move.

akaviaLab commented:
Okay.
I can move it from src/tests to tests.

However, that will cause a bit of a problem since importing/using this in cobra without the source directory (as it is installed by a package) would be difficult. Two options for this are
1) Put the function create_model in src/scripts so it is installed locally
2) Put the function create_model in cobra.util directory (for example) and have it imported into tests. Tests can import from cobra, but removing init.py means that cobra has a hard time importing from tests.

I can use Docker to check it to make sure it is present and workable, including data, but I'd like a decision from the other devs on location please, if that's possible.

Midnighter commented:
Indeed, my goal there was to have a `/tests` directory next to the `/src/cobra` directory. And I think it's actually a feature that the test suite is not installed with the source package.

The [load_model](https://cobrapy.readthedocs.io/en/latest/autoapi/cobra/io/index.html?#cobra.io.load_model) function should be the replacement for `create_test_model` in documentation and examples everywhere. I think changing examples is the real major piece of work and the reason I didn't proceed on this so far. Within the test suite itself, we should have a few pytest fixtures that provide models.

Commit:
moved to tests from src/tests

Commit:
working on test_all()

akaviaLab commented:
I think I see.
Would you like the cobra package to have some example data, so people can load_model even if there is no internet? Currently, create_test_model() has the advantage that it can work when no internet is available.

Midnighter commented:
> Would you like the cobra package to have some example data, so people can load_model even if there is no internet? Currently, create_test_model() has the advantage that it can work when no internet is available.

It's a valid concern. I'm not against packaging some models, also so that users have the exact models that examples were run with. Please note that in normal use `load_model` will cache the download locally so after first load, a connection will not be necessary. And you do need a connection to install cobrapy, of course.

akaviaLab commented:
Is there any way to setup the package so that some example models are already loaded in the cache? I don't think so, and it might not be a good idea.
However, if it is possible and a reasonable idea, one or two models can start cached.

Alternatively, a possible (but not great) idea is to add some edge cases to load_model, so if it gets "textbook", which isn't in any of the repositories, it will load it locally from cobra data. 

I'm just trying to think of users that are very new to cobrapy, and want to load some models to see behavior.

Midnighter commented:
> Is there any way to setup the package so that some example models are already loaded in the cache? I don't think so, and it might not be a good idea.
However, if it is possible and a reasonable idea, one or two models can start cached.

It can be done as a post-installation step but it's a bad idea, I think, because it assumes that the cache directory can be written to (this is not always the case).

Adding a utility function and distributing the models with the package is the best option, IMO. The only question is then if they should be in SBML or pickle format.

cdiener commented:
I think we could get rid of the pickle ones because those are a pain to maintain anyways. I think having some manually validated JSON and YAML ones would be good though to reproduce bugs etc. I agree with using load_model just have to make sure it's a new release (not backwards compatible). The test models could act like a new provider "local" or "cobrapy" similar to the bio models one.

akaviaLab commented:
Are the mini models validated?
I was thinking about getting update_pickles.py to copy some files to src/data. Set update_pickles.py  to run before release.
Then the source version is still in tests/data, and that gets modified. 

Or I could copy the mini models manually to src/data, and those would be examples.

So, load_models should have a repository like BiggModels which will be cobrapy, and that will use import_resources to import files. I think import_resources makes sure that however the users install the package, the data files are available.

Commit:
test_all() will work

Commit:
test_io directory files will work. test_io/conftest.py had errors

Commit:
inital attempt at loading from cobrapy, via cobrapy_repository.py

Commit:
some tidying of imports

Commit:
allow load_model to load locally

Commit:
modified all examples in .py file comments
gapfilling.py example now uses a model from BIGG, and requires internet

Commit:
some formatting

Commit:
fixed some errors, fixed setup.cfg, tox.ini

akaviaLab commented:
I fixed all the usages of create_test_model in the code examples.

I'd be happy to fix the python notebooks, but can you please tell me what needs to happen?
Do I just need to replace create_test_model with the appropriate call (simple text replacing like the examples)
OR do I need to manually run the Jupyter notebooks

Thank you

Commit:
constraints_objectives.ipynb source

Commit:
constraints_objectives.ipynb outputs

Commit:
constraints_objectives.ipynb code

Commit:
constraints_objectives.ipynb outputs

Commit:
cobra-cameo-merge-benchmarks-before-merge.ipynb code

Commit:
cobra-cameo-merge-benchmarks-before-merge.ipynb outputs

Commit:
deletions.ipynb code

Commit:
deletions.ipynb outputs

Commit:
configuration.ipynb code

Commit:
configuration.ipynb outputs

Commit:
add local salmonella models that doesn't have buckets of warnings

Commit:
gapfilling.ipynb code

Commit:
gapfilling.ipynb outputs

Commit:
Trying to make windows work

Commit:
cobra-cameo-merge-benchmarks.ipynb code

Commit:
cobra-cameo-merge-benchmarks.ipynb outputs

Midnighter commented:
Dear @akaviaLab,

I would love to get this one and your other works merged. I tried to send you an email in order to set up a call to discuss this and the compartments PR but it came back with access denied. Can you please write to me or head over to https://gitter.im/opencobra/cobrapy where we can talk a bit more disconnected from code.

synchon commented:
> I think changing examples is the real major piece of work and the reason I didn't proceed on this so far.

@Midnighter Can we maybe take this opportunity to have our how-to Jupyter notebooks hosted on [Binder](https://mybinder.org/) ? I believe this will let new users to play around easily with cobrapy without hitting Python and related hurdles (we have seen quite a few 😅).

>  I think having some manually validated JSON and YAML ones would be good though to reproduce bugs etc.

@cdiener Would that be a good reason to have a separate repository for them or do we continue with as it is now?

Midnighter commented:
> @Midnighter Can we maybe take this opportunity to have our how-to Jupyter notebooks hosted on [Binder](https://mybinder.org/) ? I believe this will let new users to play around easily with cobrapy without hitting Python and related hurdles (we have seen quite a few sweat_smile).

Yes, can't remember where but Binder was mentioned and approved before by everyone involved. Just™️ need to do it 😉

synchon commented:
> Yes, can't remember where but Binder was mentioned and approved before by everyone involved. Just™️ need to do it 😉

Okay then I can take a stab at it in a few days 😁

akaviaLab commented:
@Midnighter Posted on gitter
@synchon can you wait a bit before starting to work on binder? I"m working on this refactoring and discussing it with @Midnighter. Thank you

Commit:
geometric_fba.ipynb corrected.
Not running the solvers because only glpk works on my computer.

Commit:
parallel_fva.ipynb code

Commit:
parallel_fva.ipynb output

Commit:
single_gene_deletion_linear_room.ipynb changes
Not running everything since I only have glpk

Commit:
single_gene_deletion_room.ipynb changes
Not running everything since I only have glpk

Commit:
faq.ipynb fixed as much as I could fix it

Commit:
dfba.ipynb code

Commit:
dfba.ipynb outpu

Commit:
getting_started.ipynb code

Commit:
getting_started.ipynb markdown and output

Commit:
getting_started.ipynb minor error

Commit:
io.ipynb code

Commit:
io.ipynb output

Commit:
loopless.ipynb code

Commit:
loopless.ipynb output

Commit:
media.ipynb code

Commit:
media.ipynb output

Commit:
phenotype_phase_plane.ipynb code

Commit:
phenotype_phase_plane.ipynb output

Commit:
pymatbridge.ipynb code

Commit:
sampling.ipynb code

Commit:
sampling.ipynb output

Commit:
solvers.ipynb code

Commit:
simulating.ipynb code

Commit:
simulating.ipynb output

akaviaLab commented:
All the notebook files done, except for faq.ipynb, which had functions I couldn't figure out.

synchon commented:
> @synchon can you wait a bit before starting to work on binder? I"m working on this refactoring and discussing it with @Midnighter. Thank you

Yes of course. I would have waited anyway to get this PR merged. 😄

akaviaLab commented:
@synchon Great!
If you can review faq.ipynb, I'd appreciate it.

synchon commented on documentation_builder/faq.ipynb:
```diff
@@ -62,11 +62,20 @@
+    "model = cobra.io.load_model(\"iYS1720\")\n",
     "\n",
     "for metabolite in model.metabolites:\n",
     "    metabolite.id = \"test_\" + metabolite.id\n",
```
Would be nice to use f-string here to keep up with modern Python.

synchon replied:
Would be nice to use f-string here to keep up with modern Python.

synchon commented:
@akaviaLab `faq.ipynb` looks okay after your changes.

One comment about the section `How do I generate an LP file from a COBRA model?`: The solvers are now `optlang`-based and internal `cobrapy` solvers don't exist anymore, so better to remove the unnecessary sub-section. While you are at it, maybe also rename COBRA to cobrapy just to be consistent.

Also, this line `With optlang solvers, the LP formulation of a model is obtained by it's string representation. All solvers behave the same way.` has a small typo: `... obtained by its string ...`. Would be nice if you can fix that as well.

Any other enhancements, can be tackled later in a different PR later.

Commit:
fixed faq.ipynb and removed irrelevant section

akaviaLab commented:
Fixed comments. Set it up to be based on the most recent version of #1208 so it can be merged afterwards. If anything else is merged before this is done. I will rebase it.

synchon commented on documentation_builder/deletions.ipynb:
```diff
@@ -2,70 +2,57 @@
     "from time import time\n",
     "\n",
-    "import cobra.test\n",
+    "import cobra.io\n",
```
`from cobra.io import load_model`

synchon commented on documentation_builder/deletions.ipynb:
```diff
@@ -2,70 +2,57 @@
     "\n",
-    "cobra_model = cobra.test.create_test_model(\"textbook\")\n",
-    "ecoli_model = cobra.test.create_test_model(\"ecoli\")"
+    "cobra_model = cobra.io.load_model(\"textbook\")\n",
```
`cobra_model = load_model("textbook")`

synchon commented on documentation_builder/deletions.ipynb:
```diff
@@ -2,70 +2,57 @@
-    "cobra_model = cobra.test.create_test_model(\"textbook\")\n",
-    "ecoli_model = cobra.test.create_test_model(\"ecoli\")"
+    "cobra_model = cobra.io.load_model(\"textbook\")\n",
+    "ecoli_model = cobra.io.load_model(\"iJO1366\")"
```
`ecoli_model = load_model("iJO1366")`

synchon commented on documentation_builder/constraints_objectives.ipynb:
```diff
@@ -34,8 +34,8 @@
    "source": [
-    "import cobra.test\n",
-    "model = cobra.test.create_test_model('textbook')"
+    "from cobra.io.web import load_model\n",
```
`from cobra.io import load_model`

synchon commented on documentation_builder/faq.ipynb:
```diff
@@ -62,14 +62,23 @@
-    "from __future__ import print_function\n",
-    "import cobra.test\n",
-    "model = cobra.test.create_test_model()\n",
+    "import cobra.io\n",
```
`from cobra.io import load_model`

synchon commented on documentation_builder/faq.ipynb:
```diff
@@ -62,14 +62,23 @@
-    "import cobra.test\n",
-    "model = cobra.test.create_test_model()\n",
+    "import cobra.io\n",
+    "model = cobra.io.load_model(\"iYS1720\")\n",
```
`model = load_model("iYS1720")`

synchon commented on documentation_builder/faq.ipynb:
```diff
@@ -154,10 +166,10 @@
    ],
    "source": [
-    "model = cobra.test.create_test_model()\n",
+    "model = cobra.io.load_model(\"iYS1720\")\n",
```
`model = load_model("iYS1720")`

synchon commented on documentation_builder/faq.ipynb:
```diff
@@ -136,7 +148,7 @@
     "That depends on what precisely you mean by delete a gene.\n",
     "\n",
-    "If you want to simulate the model with a gene knockout, use the `cobra.manipulation.delete_model_genes` function. The effects of this function are reversed by `cobra.manipulation.undelete_model_genes`."
+    "If you want to simulate the model with a gene knockout, use the `cobra.manipulation.knock_out_model_genes` function. The effects of this function are reversed when exiting a context."
```
Better to mention usage of context early on: `... use the cobra.manipulation.knock_out_model_genes() within a context ...` or something similar.

synchon commented on documentation_builder/faq.ipynb:
```diff
@@ -154,10 +166,10 @@
     "PGI = model.reactions.get_by_id(\"PGI\")\n",
     "print(\"bounds before knockout:\", (PGI.lower_bound, PGI.upper_bound))\n",
-    "cobra.manipulation.delete_model_genes(model, [\"STM4221\"])\n",
+    "cobra.manipulation.knock_out_model_genes(model, [\"STM4221\"])\n",
```
Not sure if the import is being made before this. If not, then: `from cobra.manipulation import knock_out_model_genes` and then the necessary code.

synchon commented on documentation_builder/faq.ipynb:
```diff
@@ -199,7 +211,7 @@
    ],
    "source": [
-    "model = cobra.test.create_test_model()\n",
+    "model = cobra.io.load_model(\"iYS1720\")\n",
```
`model = load_model("iYS1720")`

synchon commented on documentation_builder/gapfilling.ipynb:
```diff
@@ -35,47 +35,43 @@
+   ],
    "source": [
-    "import cobra.test\n",
+    "import cobra.io\n",
```
`from cobra.io import load_model`

synchon commented on documentation_builder/gapfilling.ipynb:
```diff
@@ -35,47 +35,43 @@
+    "import cobra.io\n",
     "from cobra.flux_analysis import gapfill\n",
-    "model = cobra.test.create_test_model(\"salmonella\")"
+    "model = cobra.io.load_model(\"iYS1720\")"
```
`model = load_model("iYS1720")`

synchon commented on documentation_builder/getting_started.ipynb:
```diff
@@ -25,15 +25,23 @@
    "source": [
-    "from __future__ import print_function\n",
-    "\n",
     "import cobra\n",
```
If it's unused, you can remove this.

akaviaLab replied:
Might still be needed. Not sure

synchon commented on documentation_builder/getting_started.ipynb:
```diff
@@ -25,15 +25,23 @@
-    "\n",
     "import cobra\n",
-    "import cobra.test\n",
+    "import cobra.io\n",
```
`from cobra.io import load_model`

synchon commented on documentation_builder/getting_started.ipynb:
```diff
@@ -25,15 +25,23 @@
-    "# \"ecoli\" and \"salmonella\" are also valid arguments\n",
-    "model = cobra.test.create_test_model(\"textbook\")"
+    "# \"iJO1366\" and \"salmonella\" are also valid arguments\n",
+    "model = cobra.io.load_model(\"textbook\")"
```
`model = load_model("textbook")`

synchon commented on documentation_builder/getting_started.ipynb:
```diff
@@ -1175,7 +1097,7 @@
    ],
    "source": [
-    "model = cobra.test.create_test_model('textbook')\n",
+    "model = cobra.io.load_model('textbook')\n",
```
`model = load_model("textbook")`

synchon commented on documentation_builder/io.ipynb:
```diff
@@ -26,23 +26,24 @@
-    "import cobra.test\n",
-    "import os\n",
-    "from os.path import join\n",
+    "import pathlib\n",
```
`from pathlib import Path`

synchon commented on documentation_builder/io.ipynb:
```diff
@@ -26,23 +26,24 @@
+    "import logging\n",
     "\n",
-    "data_dir = cobra.test.data_dir\n",
+    "data_dir = pathlib.Path(\"\").joinpath('..', 'tests', 'data').resolve()\n",
```
I believe `tests` directory will not be distributed to users in the future and hence this would not work. Did you mean `src/data`?.

Also, you can use a modern syntax: 
```
data_dir = Path(".") / ".." / "tests" / "data"
data_dir_resolved = data_dir.resolve()
```

akaviaLab replied:
I did not move data to src/ in this PR
If you'd like me to move or copy it, let me know and I'll do it. Not all data in tests/data is needed for src/

synchon replied:
I found a new directory `src/data` during my review, so I think the move is being made. If not being moved, a new PR for the move is better.

akaviaLab replied:
Sorry, my bad. I did move example files. Revised code to src/cobra/data

synchon commented on documentation_builder/io.ipynb:
```diff
@@ -26,23 +26,24 @@
-    "textbook_model = cobra.test.create_test_model(\"textbook\")\n",
-    "ecoli_model = cobra.test.create_test_model(\"ecoli\")\n",
-    "salmonella_model = cobra.test.create_test_model(\"salmonella\")"
+    "textbook_model = cobra.io.load_model(\"textbook\")\n",
```
`textbook_model = load_model("textbook")`

synchon commented on documentation_builder/io.ipynb:
```diff
@@ -26,23 +26,24 @@
-    "ecoli_model = cobra.test.create_test_model(\"ecoli\")\n",
-    "salmonella_model = cobra.test.create_test_model(\"salmonella\")"
+    "textbook_model = cobra.io.load_model(\"textbook\")\n",
+    "ecoli_model = cobra.io.load_model(\"iJO1366\")\n",
```
`ecoli_model = load_model("iJO1366")`

synchon commented on documentation_builder/io.ipynb:
```diff
@@ -26,23 +26,24 @@
+    "textbook_model = cobra.io.load_model(\"textbook\")\n",
+    "ecoli_model = cobra.io.load_model(\"iJO1366\")\n",
+    "logging.getLogger(\"cobra.io.sbml\").setLevel(logging.ERROR) # salmonella leads to lots of warnings without this line\n",
+    "salmonella_model = cobra.io.load_model(\"salmonella\")"
```
`salmonella_model = load_model("salmonella")`

synchon commented on documentation_builder/io.ipynb:
```diff
@@ -102,15 +106,13 @@
    ],
    "source": [
-    "cobra.io.read_sbml_model(join(data_dir, \"mini_fbc2.xml\"))"
+    "cobra.io.read_sbml_model(str(data_dir.joinpath(\"mini_fbc2.xml\")))"
```
```
mini_fbc2_path = data_dir / "mini_fbc2.xml" 
read_sbml_model(str(mini_fbc2_path.resolve()))

synchon commented on documentation_builder/io.ipynb:
```diff
@@ -102,15 +106,13 @@
+   "metadata": {},
    "outputs": [],
    "source": [
     "cobra.io.write_sbml_model(textbook_model, \"test_fbc2.xml\")"
```
`write_sbml_model(textbook_model, "test_fbc2.xml")`

synchon commented on documentation_builder/io.ipynb:
```diff
@@ -168,19 +173,16 @@
    ],
    "source": [
-    "cobra.io.read_sbml_model(join(data_dir, \"mini_cobra.xml\"))"
+    "cobra.io.read_sbml_model(str(data_dir.joinpath(\"mini_cobra.xml\")))"
```
```
mini_cobra_path = data_dir / "mini_cobra.xml" 
read_sbml_model(str(mini_cobra_path.resolve()))
```

synchon commented on documentation_builder/io.ipynb:
```diff
@@ -168,19 +173,16 @@
    "source": [
-    "cobra.io.write_sbml_model(\n",
-    "    textbook_model, \"test_cobra.xml\", use_fbc_package=False)"
+    "cobra.io.write_sbml_model(textbook_model, \"test_cobra.xml\")"
```
`write_sbml_model(textbook_model, "test_cobra.xml")`

synchon commented on documentation_builder/io.ipynb:
```diff
@@ -238,15 +243,13 @@
    ],
    "source": [
-    "cobra.io.load_json_model(join(data_dir, \"mini.json\"))"
+    "cobra.io.load_json_model(str(data_dir.joinpath(\"mini.json\")))"
```
```
mini_json_path = data_dir / "mini.json" 
load_json_model(str(mini_json_path.resolve()))
```

synchon commented on documentation_builder/io.ipynb:
```diff
@@ -238,15 +243,13 @@
+   "metadata": {},
    "outputs": [],
    "source": [
     "cobra.io.save_json_model(textbook_model, \"test.json\")"
```
`save_json_model(textbook_model, "test.json")`

synchon commented on documentation_builder/io.ipynb:
```diff
@@ -307,15 +313,13 @@
    ],
    "source": [
-    "cobra.io.load_yaml_model(join(data_dir, \"mini.yml\"))"
+    "cobra.io.load_yaml_model(str(data_dir.joinpath(\"mini.yml\")))"
```
```
mini_yml_path = data_dir / "mini.yml"
load_yaml_model(str(mini_yml_path.resolve()))
```

synchon commented on documentation_builder/io.ipynb:
```diff
@@ -379,7 +386,7 @@
    "source": [
     "cobra.io.load_matlab_model(\n",
-    "    join(data_dir, \"mini.mat\"), variable_name=\"mini_textbook\")"
+    "    str(data_dir.joinpath(\"mini.mat\")), variable_name=\"mini_textbook\")"
```
```
mini_mat_path = data_dir / "mini.mat"
load_matlab_model(
    str(mini_mat_path.resolve()),
    variable_name="mini_textbook",
)

synchon commented on documentation_builder/io.ipynb:
```diff
@@ -430,7 +440,7 @@
    ],
    "source": [
-    "cobra.io.load_matlab_model(join(data_dir, \"mini.mat\"))"
+    "cobra.io.load_matlab_model(str(data_dir.joinpath(\"mini.mat\")))"
```
```
mini_mat_path = data_dir / "mini.mat"
load_matlab_model(str(mini_mat_path.resolve()))
```

synchon commented on documentation_builder/io.ipynb:
```diff
@@ -26,23 +26,24 @@
-    "import os\n",
-    "from os.path import join\n",
+    "import pathlib\n",
+    "import cobra\n",
```
Import only the necessary functions from `cobra.io`

synchon commented on documentation_builder/phenotype_phase_plane.ipynb:
```diff
@@ -20,12 +20,22 @@
+   ],
    "source": [
-    "import cobra.test\n",
+    "import cobra.io\n",
```
`from cobra.io import load_model`

synchon commented on documentation_builder/phenotype_phase_plane.ipynb:
```diff
@@ -20,12 +20,22 @@
     "from cobra.flux_analysis import production_envelope\n",
     "\n",
-    "model = cobra.test.create_test_model(\"textbook\")"
+    "model = cobra.io.load_model(\"textbook\")"
```
`model = load_model("textbook")`

synchon commented on documentation_builder/pymatbridge.ipynb:
```diff
@@ -40,8 +36,8 @@
    "source": [
-    "import cobra.test\n",
-    "m = cobra.test.create_test_model(\"textbook\")"
+    "import cobra.io\n",
```
`from cobra.io import load_model`

synchon commented on documentation_builder/pymatbridge.ipynb:
```diff
@@ -40,8 +36,8 @@
-    "import cobra.test\n",
-    "m = cobra.test.create_test_model(\"textbook\")"
+    "import cobra.io\n",
+    "m = cobra.io.load_model(\"textbook\")"
```
`m = load_model("textbook")`

synchon commented on documentation_builder/simulating.ipynb:
```diff
@@ -18,10 +18,20 @@
    "source": [
-    "import cobra.test\n",
-    "model = cobra.test.create_test_model(\"textbook\")"
+    "import cobra.io\n",
```
`from cobra.io import load_model`

synchon commented on documentation_builder/simulating.ipynb:
```diff
@@ -18,10 +18,20 @@
-    "import cobra.test\n",
-    "model = cobra.test.create_test_model(\"textbook\")"
+    "import cobra.io\n",
+    "model = cobra.io.load_model(\"textbook\")"
```
`model = load_model("textbook")`

synchon commented on documentation_builder/solvers.ipynb:
```diff
@@ -24,25 +21,17 @@
    "source": [
-    "import cobra.test\n",
-    "model = cobra.test.create_test_model('textbook')"
+    "import cobra.io\n",
```
`from cobra.io import load_model`

synchon commented on documentation_builder/solvers.ipynb:
```diff
@@ -24,25 +21,17 @@
-    "import cobra.test\n",
-    "model = cobra.test.create_test_model('textbook')"
+    "import cobra.io\n",
+    "model = cobra.io.load_model('textbook')"
```
`model = load_model("textbook")`

synchon commented on src/cobra/core/dictlist.py:
```diff
@@ -134,8 +134,8 @@ def query(
         --------
-        >>> import cobra.test
-        >>> model = cobra.test.create_test_model('textbook')
+        >>> import cobra.io
```
`from cobra.io import load_model`

synchon commented on src/cobra/core/dictlist.py:
```diff
@@ -134,8 +134,8 @@ def query(
-        >>> import cobra.test
-        >>> model = cobra.test.create_test_model('textbook')
+        >>> import cobra.io
+        >>> model = cobra.io.load_model('textbook')
```
`model = load_model("textbook")`

synchon commented on src/cobra/core/metabolite.py:
```diff
@@ -241,8 +241,8 @@ def shadow_price(self) -> float:
         >>> import cobra
-        >>> import cobra.test
-        >>> model = cobra.test.create_test_model("textbook")
+        >>> import cobra.io
```
`from cobra.io import load_model`

synchon commented on src/cobra/core/metabolite.py:
```diff
@@ -241,8 +241,8 @@ def shadow_price(self) -> float:
         Examples
         --------
         >>> import cobra
```
Not needed.

synchon commented on src/cobra/core/metabolite.py:
```diff
@@ -241,8 +241,8 @@ def shadow_price(self) -> float:
-        >>> import cobra.test
-        >>> model = cobra.test.create_test_model("textbook")
+        >>> import cobra.io
+        >>> model = cobra.io.load_model("textbook")
```
`model = load_model("textbook")`

synchon commented on src/cobra/core/model.py:
```diff
@@ -129,10 +126,9 @@ def solver(self):
-        >>> model = cobra.test.create_test_model("textbook")
-        >>> new = model.problem.Constraint(model.objective.expression,
-        >>> lb=0.99)
+        >>> import cobra.io
```
`from cobra.io import load_model`

synchon commented on src/cobra/core/model.py:
```diff
@@ -129,10 +126,9 @@ def solver(self):
-        >>> new = model.problem.Constraint(model.objective.expression,
-        >>> lb=0.99)
+        >>> import cobra.io
+        >>> model = cobra.io.load_model("textbook")
```
`model = load_model("textbook")`

synchon commented on src/cobra/core/model.py:
```diff
@@ -220,9 +216,10 @@ def compartments(self, value):
         --------
-        >>> import cobra.test
-        >>> model = cobra.test.create_test_model("textbook")
+        >>> import cobra.io
```
`from cobra.io import load_model`

synchon commented on src/cobra/core/model.py:
```diff
@@ -220,9 +216,10 @@ def compartments(self, value):
-        >>> import cobra.test
-        >>> model = cobra.test.create_test_model("textbook")
+        >>> import cobra.io
+        >>> model = cobra.io.load_model("textbook")
```
`model = load_model("textbook")`

synchon commented on src/cobra/core/model.py:
```diff
@@ -583,8 +580,8 @@ def add_boundary(
         --------
-        >>> import cobra.test
-        >>> model = cobra.test.create_test_model("textbook")
+        >>> import cobra.io
```
`from cobra.io import load_model`

synchon commented on src/cobra/core/model.py:
```diff
@@ -583,8 +580,8 @@ def add_boundary(
-        >>> import cobra.test
-        >>> model = cobra.test.create_test_model("textbook")
+        >>> import cobra.io
+        >>> model = cobra.io.load_model("textbook")
```
`model = load_model("textbook")`

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -485,8 +485,8 @@ def flux(self) -> float:
         --------
-        >>> import cobra.test
-        >>> model = cobra.test.create_test_model("textbook")
+        >>> import cobra.io
```
`from cobra.io import load_model`

synchon commented on src/cobra/core/reaction.py:
```diff
@@ -485,8 +485,8 @@ def flux(self) -> float:
-        >>> import cobra.test
-        >>> model = cobra.test.create_test_model("textbook")
+        >>> import cobra.io
+        >>> model = cobra.io.load_model("textbook")
```
`model = load_model("textbook")`

synchon commented on src/cobra/flux_analysis/gapfilling.py:
```diff
@@ -385,10 +385,10 @@ def gapfill(
     Examples
     --------
-    >>> import cobra.test as ct
+    >>> import cobra.io as cio
```
`from cobra.io import load_model`

synchon commented on src/cobra/flux_analysis/gapfilling.py:
```diff
@@ -385,10 +385,10 @@ def gapfill(
     >>> from cobra import Model
     >>> from cobra.flux_analysis import gapfill
-    >>> model = ct.create_test_model("salmonella")
+    >>> model = cio.load_model("iYS1720")
```
`model = load_model("iYS1720")`

synchon commented on src/cobra/flux_analysis/phenotype_phase_plane.py:
```diff
@@ -81,9 +81,9 @@ def production_envelope(
     Examples
     --------
-    >>> import cobra.test
+    >>> import cobra.io
```
`from cobra.io import load_model`

synchon commented on src/cobra/flux_analysis/phenotype_phase_plane.py:
```diff
@@ -81,9 +81,9 @@ def production_envelope(
+    >>> import cobra.io
     >>> from cobra.flux_analysis import production_envelope
-    >>> model = cobra.test.create_test_model("textbook")
+    >>> model = cobra.io.load_model("textbook")
```
`model = load_model("textbook")`

synchon commented on src/cobra/manipulation/delete.py:
```diff
@@ -185,62 +108,39 @@ def delete_model_genes(
     cumulative_deletions: bool, optional
         If True, then any previous deletions will be maintained in the
-        model (default True).
+        model (default True). Unused, ignored.
```
Any reason for adding this?

akaviaLab replied:
This is part of #1208 which will hopefully be merged soon. Wanted to minimize conflicts.

synchon replied:
I think I found a deprecation notice in the docstring, during my review. Was a long review so don't remember clearly. If the notice is there, you can remove this.

akaviaLab replied:
Let's leave this PR alone for now. If you want to review #1208, I will merge all changes of it into the tests. Otherwise, I'm not going to change it here.

synchon commented on tests/test_io/test_pickle.py:
```diff
@@ -23,7 +21,7 @@
 
 @pytest.mark.parametrize("load_function", [load, cload])
-def test_read_pickle(data_directory, mini_model, load_function):
+def test_read_pickle(compare_models, data_directory, mini_model, load_function):
```
Please add the necessary type annotations.

synchon commented on benchmarks/cobra-cameo-merge-benchmarks-before-merge.ipynb:
```diff
@@ -2,32 +2,28 @@
     "import cobra\n",
     "from cobra.io import read_sbml_model\n",
-    "from cobra.test import create_test_model\n",
+    "from cobra.io import load_model\n",
```
This import and the previous import can be merged.

synchon commented on benchmarks/cobra-cameo-merge-benchmarks-before-merge.ipynb:
```diff
@@ -2,32 +2,28 @@
+   "metadata": {},
    "outputs": [],
    "source": [
     "import cobra\n",
```
Remove if not needed.

akaviaLab replied:
Still needed.

synchon replied:
Could you please point out where?

synchon commented on benchmarks/cobra-cameo-merge-benchmarks.ipynb:
```diff
@@ -3,172 +3,23 @@
+   "metadata": {},
+   "outputs": [],
    "source": [
     "import cobra\n",
```
Remove if not needed.

akaviaLab replied:
Needed for cobra.__path__ and cobra.__version__ below. Not going to change them in this PR.

synchon replied:
I see, okay.

synchon reviewed (changes requested):
Looks good. Thanks for the effort!

akaviaLab reviewed (commented):
Fixed your comments. When the review is done, I will merge the changes to the mat file and resolve conflict - not doing it now, since it will break conversations.

Commit:
fixed according to comments

synchon commented on tests/test_io/test_json.py:
```diff
@@ -4,6 +4,7 @@
 from os.path import join
 from typing import Any, Callable, Dict, Union
 
+import py.test
```
Not sure about this import. Any reason for using this?

akaviaLab replied:
Because tmpdir is neither "str" nor "Path.tempdir" nor anything else. It is actually py.test, and I wanted the typing to be correct if doing it.

synchon replied:
Okay fair enough. We can revisit later if needed.

Commit:
changes in io from tests/data to src/cobra/data

akaviaLab commented:
If the conversations are resolved as far as you are concerned, I will resolve the conflicts of the mat related files. Let me know if that's happened.

synchon commented on documentation_builder/io.ipynb:
```diff
@@ -35,7 +35,8 @@
     "\n",
-    "data_dir = pathlib.Path(\"\").joinpath('..', 'tests', 'data').resolve()\n",
+    "data_dir = Path(\".\") / \"..\" / \"src\" / \"cobra\" / \"data\"\n",
+    "data_dir_resolved = data_dir.resolve()\n",
```
I don't remember if this is being used anywhere. I wrote it in the beginning of the review, so if it's not being used, you can remove it. My bad.

akaviaLab replied:
We need to resolve the dir othewise it might now work. Fixed it now.

Commit:
io.ipynb should work now

synchon commented:
I have no further comments on this. Make sure notebooks are working properly. You can resolve the conflicts and push the changes. Hopefully, the CI/CD works fine. Just a bit of extra caution due to significant changes.

akaviaLab commented:
All my work on this was opening the notebooks and running them manually to
make sure they work. The changes shouldn't have broken them.

On Thu, Apr 28, 2022 at 5:08 PM Synchon Mandal ***@***.***>
wrote:

> I have no further comments on this. Make sure notebooks are working
> properly. You can resolve the conflicts and push the changes. Hopefully,
> the CI/CD works fine. Just a bit of extra caution due to significant
> changes.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/opencobra/cobrapy/pull/1191#issuecomment-1112655840>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACQYYZT2BXRWEN4OARGD2GTVHL45XANCNFSM5ROKAPJQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

synchon commented:
Perfect! Let's see how the CI/CD behaves.

Commit:
fixed py.path instead of py.test

Commit:
delet pymatbridge.ipynb

Commit:
fixed more merge problems

Commit:
fixed changes in compare_models

Midnighter commented:
Need to find a good way to review this. Was the move into a separate tests directory in one particular commit? I'd like to ignore that one.

synchon commented:
@Midnighter I believe `2ee0dcbfa260199cc7baac2c9e504fa6a43bb670` was the one.

Midnighter commented:
Okay, @akaviaLab can you please rebase on the latest devel one last time? I won't merge any other work until this one is in but the commit history here is now pretty messy.

akaviaLab commented:
@Midnighter  - would you like me to squash the commits of the notebooks? I had one commit for each notebook code and one commit for each notebook output.
I can squash them to one commit per notebook if that will help.

Midnighter commented:
> @Midnighter - would you like me to squash the commits of the notebooks? I had one commit for each notebook code and one commit for each notebook output. I can squash them to one commit per notebook if that will help.

No, that actually sounds good to me.

Midnighter commented:
I see now that you moved things to `/src/tests` rather than `/tests`. Would it have a huge impact on the other commits if you changed that now?

akaviaLab commented:
Check a few commits down
First I moved to stc/tests  and then I moved to /tests

On Fri., Apr. 29, 2022, 6:25 a.m. Moritz E. Beber, ***@***.***>
wrote:

> I see now that you moved things to /src/tests rather than /tests. Would
> it have a huge impact on the other commits if you changed that now?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/opencobra/cobrapy/pull/1191#issuecomment-1113150046>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACQYYZXZSTSCVSC3YRCKP7DVHO2I5ANCNFSM5ROKAPJQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

Midnighter commented on setup.cfg:
```diff
@@ -79,8 +79,8 @@ where = src
 cobra =
-	io/*.json
-	test/data/*
+    data/*
```
I would like to reduce the models that we distribute with the package. Most of those are only really useful for testing and should be part of the repository but not the distributed Python package, I think.

Midnighter replied:
Essentially, keep all SBML models for the examples and the others should be moved into a `tests/data` directory or so.

akaviaLab replied:
Hmm. I kept yaml, json, pickle and mat in the mini model format, just so we can show what they're supposed to look like.
I can definitely delete iJO1366.pickle.
As for the mini models - I'll agree to what you want.

Midnighter replied:
Fair enough, they hardly hurt. Then just remove the pickle file, please.

Midnighter commented on src/cobra/io/web/cobrapy_repository.py:
```diff
@@ -0,0 +1,53 @@
+        """
+        cobra_data_dir = importlib_resources.files("cobra.data")
+        model_file = cobra_data_dir.joinpath(f"{model_id}.xml.gz")
+        return model_file.read_bytes()
```
I'd suggest the following. Add a `__init__.py` module to `cobra.data` and import it at the top of this module. Then

```suggestion
        return importlib_resources.files(cobra.data).joinpath(f"{model_id}.xml.gz").read_bytes()
```

Midnighter commented on release-notes/next-release.md:
```diff
@@ -22,6 +22,9 @@ fail more comprehensibly.
 ## Other
 
+Moved tests to a different directory, from `src/cobra/test` 
+to `src/tests`
```
```suggestion
to `tests`
```

Midnighter commented on src/cobra/io/web/load.py:
```diff
@@ -26,12 +27,18 @@
     Download an SBML model from a remote repository.
 
+    It will try the local repository first.
+
```
This statement is only true with default arguments. I'd remove it.

```suggestion
```

Midnighter commented on tests/conftest.py:
```diff
@@ -1,18 +1,67 @@
+
+cobra_directory = abspath(join(dirname(abspath(__file__)), ".."))
+cobra_location = abspath(join(cobra_directory, "src"))
+data_dir = join(cobra_directory, "tests", "data", "")
```
At the top, `from pathlib import Path`, then as mentioned before, move the data directory from `src/cobra/data` to `tests/data` (with the exception of the SBML files that should remain there or be in both places).

```suggestion
data_dir = Path(__file__).parent / "data"
```

Midnighter commented on tests/conftest.py:
```diff
@@ -1,18 +1,67 @@
+        model_name = salmonella_pickle
+    with open(model_name, "rb") as infile:
+        return _load(infile)
+
+
+def test_all(args=None):
+    """Alias for running all unit-tests on installed cobra."""
+    if pytest:
+        args = args if args else []
+
+        return pytest.main(
+            [
+                str(abspath(join(cobra_directory, "tests"))),
+                "--benchmark-skip",
+                "-v",
+                "-rs",
+            ]
+            + args
+        )
+    else:
+        raise ImportError(
+            "missing package pytest and pytest_benchmark required for testing"
+        )
```
It seems to me that this is no longer needed.

akaviaLab replied:
Yeah. I wasn't sure about this. Let's deal with everything and then try removing it.

Midnighter commented on tests/test_io/conftest.py:
```diff
@@ -0,0 +1,136 @@
+
+
+@pytest.fixture(scope="module")
+def mini_model(data_directory: str):
```
```suggestion
def mini_model(data_directory: Path) -> Model:
```

Midnighter commented on tests/conftest.py:
```diff
@@ -1,18 +1,67 @@
+
+    """
+    if model_name == "ecoli":
+        ecoli_sbml = join(data_dir, "iJO1366.xml.gz")
```
With `data_dir` being a `Path` you might have to change these to:

```suggestion
        ecoli_sbml = str(data_dir / "iJO1366.xml.gz")
```

Midnighter commented on tests/conftest.py:
```diff
@@ -1,18 +1,67 @@
+data_dir = join(cobra_directory, "tests", "data", "")
+
+
+def create_test_model(model_name="salmonella"):
```
```suggestion
def create_test_model(model_name="salmonella") -> Model:
```

Midnighter commented on tests/test_io/conftest.py:
```diff
@@ -0,0 +1,136 @@
+@pytest.fixture(scope="module")
+def mini_model(data_directory: str):
+    """Fixture for mini model."""
+    with open(join(data_directory, "mini.pickle"), "rb") as infile:
```
```suggestion
    with (data_directory / "mini.pickle").open(mode="rb") as infile:
```

Midnighter commented on tox.ini:
```diff
@@ -95,15 +95,15 @@ source =
 parallel = true
 omit =
-    */cobra/test/*
+    tests/*
```
Coverage will no longer be measured on the tests themselves. This can be removed entirely.

Midnighter commented on tox.ini:
```diff
@@ -95,15 +95,15 @@ source =
 precision = 2
 omit =
-    */cobra/test/*
+    tests/*
```
As above, we don't need to omit these.

Midnighter commented on tests/test_io/test_json.py:
```diff
@@ -2,20 +2,20 @@
-from typing import Any, Dict, Union
+from typing import Any, Callable, Dict, Union
 
+import py.path
```
Let's upgrade the fixture used, then this won't be needed.

akaviaLab replied:
Can you please give me a bit more detail about what you mean with this comment?

akaviaLab replied:
Are you talking about my compare function that is pending #1206

Midnighter replied:
No, I mean using `tmp_path` (type `Path`) instead of `tmpdir`.

Midnighter commented on tests/test_io/test_json.py:
```diff
@@ -30,14 +30,16 @@ def test_validate_json(
 
-def test_load_json_model(data_directory: str, mini_model: Model) -> None:
+def test_load_json_model(
+    compare_models: Callable, data_directory: str, mini_model: Model
```
```suggestion
    compare_models: Callable, data_directory: Path, mini_model: Model
```

Midnighter commented on tests/test_io/test_json.py:
```diff
@@ -30,14 +30,16 @@ def test_validate_json(
+    compare_models: Callable, data_directory: str, mini_model: Model
+) -> None:
     """Test the reading of JSON model."""
     json_model = cio.load_json_model(join(data_directory, "mini.json"))
```
```suggestion
    json_model = cio.load_json_model(data_directory / "mini.json")
```

Midnighter commented on tests/test_io/test_json.py:
```diff
@@ -30,14 +30,16 @@ def test_validate_json(
 
 def test_save_json_model(
-    tmpdir: str,
+    tmpdir: "py.path.local",
```
```suggestion
    tmp_path: Path,
```

Midnighter replied:
This means that below has to be changed, too.

```py
output_file = tmp_path / "mini.json"
```

akaviaLab replied:
Wouldn't that fail unless we modified test_io_order.py as well?
Specifically
`@pytest.fixture(scope="module")
def tmp_path(tmpdir_factory):
    return str(tmpdir_factory.mktemp("model_order"))`

Midnighter replied:
I meant the pytest fixture `tmp_path`, I'd forgotten that we have a local one as well 😕 Should be replaced with `tmp_path_factory` and our local fixture renamed.

Midnighter commented on tests/test_io/test_mat.py:
```diff
@@ -17,6 +16,8 @@
 
 
 if TYPE_CHECKING:
+    from typing import Callable
```
This is redundant with above imports.

Midnighter commented on tests/test_io/test_mat.py:
```diff
@@ -37,7 +38,10 @@ def raven_model(data_directory: str) -> "Model":
 def test_load_matlab_model(
-    data_directory: str, mini_model: "Model", raven_model: "Model"
+    compare_models: Callable,
+    data_directory: str,
```
```suggestion
    data_directory: Path,
```

Midnighter commented on tests/test_io/test_mat.py:
```diff
@@ -80,7 +84,7 @@ def test_large_bounds(tmpdir: "py.path.local", model: "Model") -> None:
 @pytest.mark.skipif(scipy is None, reason="scipy unavailable")
 def test_read_rewrite_matlab_model(
-    tmpdir: "py.path.local", data_directory: str
+    compare_models: Callable, tmpdir: "py.path.local", data_directory: str
```
Again, I'd modify the test to use
```suggestion
    compare_models: Callable, tmp_path: Path, data_directory: Path
```

Midnighter commented on tests/test_io/test_mat.py:
```diff
@@ -153,7 +157,10 @@ def _fix_xml_annotation_to_identifiers(model: "Model") -> None:
-    data_directory: str, tmpdir: "py.path.local", xml_file: str
+    compare_models: Callable,
+    data_directory: str,
+    tmpdir: "py.path.local",
```
```suggestion
    data_directory: Path,
    tmp_path: Path,
```

Midnighter commented on tests/test_io/test_mat.py:
```diff
@@ -181,7 +188,7 @@ def test_fail_on_problematic_compartments(data_directory: str) -> None:
 @pytest.mark.skipif(scipy is None, reason="scipy unavailable")
 def test_mat_model_with_long_compartment_ids(
-    data_directory: str, tmpdir: "py.path.local"
+    compare_models: Callable, data_directory: str, tmpdir: "py.path.local"
```
As before

Midnighter commented on tests/test_io/test_mat.py:
```diff
@@ -211,7 +218,9 @@ def test_mat_model_with_long_compartment_ids(
 @pytest.mark.skipif(scipy is None, reason="scipy unavailable")
-def test_mat_model_with_no_genes(data_directory: str, tmpdir: "py.path.local") -> None:
+def test_mat_model_with_no_genes(
+    compare_models: Callable, data_directory: str, tmpdir: "py.path.local"
```
As before

Midnighter commented on tests/test_io/test_mat.py:
```diff
@@ -224,7 +233,7 @@ def test_mat_model_with_no_genes(data_directory: str, tmpdir: "py.path.local") -
 
 @pytest.mark.skipif(scipy is None, reason="scipy unavailable")
-def test_mat_model_wrong_caps(data_directory: str) -> None:
+def test_mat_model_wrong_caps(compare_models: Callable, data_directory: str) -> None:
```
```suggestion
def test_mat_model_wrong_caps(compare_models: Callable, data_directory: Path) -> None:
```

Midnighter commented on tests/test_io/test_pickle.py:
```diff
@@ -1,29 +1,22 @@
 from pickle import dump, load
+from typing import Callable
 
+import py.path
```
```suggestion
```

Midnighter commented on tests/test_io/test_pickle.py:
```diff
@@ -1,29 +1,22 @@
+@pytest.mark.parametrize("load_function", [load])
+def test_read_pickle(
+    compare_models: Callable,
+    data_directory: str,
```
```suggestion
    data_directory: Path,
```

Midnighter commented on tests/test_io/test_pickle.py:
```diff
@@ -34,8 +27,10 @@ def test_read_pickle(data_directory, mini_model, load_function):
-def test_write_pickle(tmpdir, mini_model, dump_function):
+@pytest.mark.parametrize("dump_function", [dump])
+def test_write_pickle(
+    tmpdir: "py.path.local", mini_model: Model, dump_function: Callable
```
```suggestion
    tmp_path: Path, mini_model: Model, dump_function: Callable
```

Midnighter commented on tests/test_io/test_yaml.py:
```diff
@@ -2,23 +2,26 @@
 from os.path import join
+from typing import Callable
 
+import py.path
```
```suggestion
```

Midnighter commented on tests/test_io/test_yaml.py:
```diff
@@ -2,23 +2,26 @@
 
-def test_load_yaml_model(data_directory: str, mini_model: Model) -> None:
+def test_load_yaml_model(
+    compare_models: Callable, data_directory: str, mini_model: Model
```
```suggestion
    compare_models: Callable, data_directory: Path, mini_model: Model
```

Midnighter commented on tests/test_io/test_yaml.py:
```diff
@@ -2,23 +2,26 @@
 
 @pytest.mark.xfail(reason="schema outdated")
-def test_save_yaml_model(tmpdir: str, mini_model: Model) -> None:
+def test_save_yaml_model(tmpdir: "py.path.local", mini_model: Model) -> None:
```
```suggestion
def test_save_yaml_model(tmp_path: Path, mini_model: Model) -> None:
```

Midnighter reviewed (changes requested):
Phew, thank you so much for tackling this 🙂 I have a few more suggestions but overall it looks very good.

One bigger point is the exact models that we want to distribute with COBRApy. Seems like only the SBML files are needed there but maybe I overlooked something.

We should also look at including [jupytext](https://jupytext.readthedocs.io/) because those notebooks diffs are just painful. That's not for this PR, though.

akaviaLab committed:
Apply suggestions from code review - mostly path

Co-authored-by: Moritz E. Beber <midnighter@posteo.net>

Commit:
deleted iOJ1366.pickle
using tmp_path instead of py.test (need to remove os.join in later PR)

Commit:
isort  black and flake8

Midnighter commented:
Looks good to me now. As soon as @synchon is happy, we can merge it. Cheers @akaviaLab

akaviaLab commented:
Might need to remove test_all
I was waiting in that

On Fri., Apr. 29, 2022, 10:30 a.m. Moritz E. Beber, <
***@***.***> wrote:

> Looks good to me now. As soon as @synchon <https://github.com/synchon> is
> happy, we can merge it. Cheers @akaviaLab <https://github.com/akaviaLab>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/opencobra/cobrapy/pull/1191#issuecomment-1113392272>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACQYYZTT5UFVEO4PGEAPYWTVHPW7DANCNFSM5ROKAPJQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

synchon reviewed (approved):
Perfect! Let's merge this. 🎉

synchon commented:
@Midnighter can we make a release after merging this PR?

Midnighter commented:
> @Midnighter can we make a release after merging this PR?

Sure

> Might need to remove test_all

Yep, please do.

Commit:
removed test_all and pytest_addoption which seem irrelevant
